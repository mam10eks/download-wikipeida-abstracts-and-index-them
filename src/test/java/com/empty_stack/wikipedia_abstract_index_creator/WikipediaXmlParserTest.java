package com.empty_stack.wikipedia_abstract_index_creator;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

import org.apache.lucene.document.Document;
import org.apache.lucene.index.IndexWriter;
import org.apache.lucene.index.IndexableField;
import org.approvaltests.Approvals;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Matchers;
import org.mockito.Mockito;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;

import lombok.SneakyThrows;

/**
 * 
 * @author Maik Fr√∂be
 *
 */
public class WikipediaXmlParserTest
{
	private WikipediaXmlParser xmlParser;
	
	private List<Iterable<? extends IndexableField>> documentsAddedToIndex;
	
	@Before
	@SneakyThrows
	public void prepareTest()
	{
		documentsAddedToIndex = new ArrayList<>();
		IndexWriter mockedIndexWriter = Mockito.mock(IndexWriter.class);
		Mockito.when(mockedIndexWriter.addDocument(Matchers.any())).thenAnswer(new Answer<Long>()
		{
			@Override
			public Long answer(InvocationOnMock invocation) throws Throwable
			{
				if(invocation.getArguments() == null || invocation.getArguments().length != 1 || 
					!(invocation.getArguments()[0] instanceof Document))
				{
					throw new RuntimeException("It is assumed that the mocked index writer only indexes Documents");
				}
				
				documentsAddedToIndex.add((Document)invocation.getArguments()[0]);
				return 0l;
			}
		});
		
		xmlParser = new WikipediaXmlParser(mockedIndexWriter);
	}
	
	private static InputStream testResourceAsInputStream(String resourceName)
	{
		return WikipediaXmlParserTest.class.getResourceAsStream(resourceName);
	}
	
	@Test
	public void approveInputWithAbstractAndUrlAndTitle()
	{
		xmlParser.parseXmlWhileAddingToIndex(testResourceAsInputStream("InputWithAbstractAndUrlAndTitle.xml"));
		
		Approvals.verify(documentsAddedToIndex);
	}
	
	@Test
	public void approveInputWithUrl()
	{
		xmlParser.parseXmlWhileAddingToIndex(testResourceAsInputStream("InputWithUrlAndTitle.xml"));
		
		Approvals.verify(documentsAddedToIndex);
	}
	
	@Test(expected=RuntimeException.class)
	public void checkThatInputWithoutTitleThrowsAnException()
	{
		xmlParser.parseXmlWhileAddingToIndex(testResourceAsInputStream("InputWithoutTitle.xml"));
	}
	
	@Test(expected=RuntimeException.class)
	public void checkThatInputWithoutUrlThrowsAnException()
	{
		xmlParser.parseXmlWhileAddingToIndex(testResourceAsInputStream("InputWithoutUrl.xml"));
	}
}
