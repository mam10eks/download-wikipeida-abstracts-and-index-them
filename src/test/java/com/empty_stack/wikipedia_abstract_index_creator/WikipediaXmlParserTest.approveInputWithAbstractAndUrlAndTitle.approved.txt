[Document<stored,indexed,tokenized<title:Wikipedia: Alan Smithee> stored,indexed,tokenized<link:https://de.wikipedia.org/wiki/Alan_Smithee> stored,indexed,tokenized<content:Alan Smithee steht als Pseudonym für einen fiktiven Regisseur, der Filme verantwortet, bei denen der eigentliche Regisseur seinen Namen nicht mit dem Werk in Verbindung gebracht haben möchte. Von 1968 bis 2000 wurde es von der Directors Guild of America (DGA) für solche Situationen empfohlen, seither ist es Thomas Lee.>>, Document<stored,indexed,tokenized<title:Wikipedia: Actinium> stored,indexed,tokenized<link:https://de.wikipedia.org/wiki/Actinium> stored,indexed,tokenized<content:| GHS-Signalwort  =>>, Document<stored,indexed,tokenized<title:Wikipedia: Ang Lee> stored,indexed,tokenized<link:https://de.wikipedia.org/wiki/Ang_Lee> stored,indexed,tokenized<content:Ang Lee (; * 23. Oktober 1954 in Pingtung, Taiwan) ist ein US-amerikanisch-taiwanischer Filmregisseur, Drehbuchautor und Produzent.>>]